src/hooks/useSelectRow.js

```diff
@@ -0,0 +1,56 @@
+import { useMemo, useState, useEffect, useCallback } from 'react'
+
+export const useSelectRow = props => {
+  const { rows, data } = props
+  const [selected, setSelected] = useState([])
+
+  useEffect(() => {
+    setSelected([])
+  }, [data])
+
+  const toggleSelectAll = forcedState => {
+    return setSelected(state => {
+      if (state.length === rows.length || forcedState === false) {
+        return []
+      }
+
+      return rows.map(row => row.index)
+    })
+  }
+
+  const toggleSelected = useCallback(
+    index => {
+      return setSelected(state => {
+        const isSelected = selected.includes(index)
+
+        if (isSelected) {
+          return selected.filter(selectedIndex => selectedIndex !== index)
+        }
+
+        return [...state, index]
+      })
+    },
+    [selected]
+  )
+
+  const mutatedRows = useMemo(() => {
+    rows.forEach(row => {
+      row.isSelected = selected.includes(row.index)
+      row.toggleSelected = () => toggleSelected(row.index)
+    })
+
+    return rows
+  }, [rows, selected, toggleSelected])
```

Comment:
I know this is necessary, but something here is smelly to me. Could be:
- That forEach loop could get expensive on large datasets. I wonder if we could hook into an existing loop over the data?
- Using useMemo, but *mutating* the rows, instead of replacing them with new objects. Usually I favor immutability, but maybe this is the right way?
- Every time the selected state changes, it will loop over all of the rows. Again, this could get expensive and slow for large datasets.

Reply:
>That forEach loop could get expensive on large datasets. I wonder if we could hook into an existing loop over the data?
>Every time the selected state changes, it will loop over all of the rows. Again, this could get expensive and slow for large datasets.
Is that something the `api.hooks.rows` could solve?
If I understand the `api.hooks` thing well, you designed that exactly for that purpose, or I'm mistaken?

>Using useMemo, but mutating the rows, instead of replacing them with new objects. Usually I favor immutability, but maybe this is the right way?
Yeah I agree absolutely, tbh I think I just followed an example from another hook (I think it was `usePagination` but anyway).